File: randomCs.ts
import * as Random from './random';
import { postError } from './jsCommon';

inlets = 2;
outlets = 2;

var maxLength: number = 64;

var order: number[] = [0, maxLength - 1]

search:
for (let denominator = 2; denominator < 100; denominator++) {
    for (let numerator = 1; numerator < denominator; numerator++) {
        const fraction = numerator / denominator;
        const number = Math.round(fraction * maxLength);
        if (order.indexOf(number) === -1) {
            order.push(number);
            if (order.length === maxLength) {
                break search;
            }
        }
    }
}

let seed: number = 0;
let take: number = 16;

let allAs: number[] = [];
let allBs: number[] = [];

function msg_int(b: number) {
    try {
        switch (inlet) {
            case 0:
                take = b;
                break;
            case 1:
                seed = b;
                regenerate();
                break;
        }
        bang();
    } catch (e) {
        postError(e);
    }
}

function regenerate() {
    allAs = [];
    allBs = [];
    let rng = new Random.MT(seed);
    for (let i = 0; i < maxLength; i++) {
        do {
            var b = 2 * (rng.normal() - 0.5);
            var c = 2 * (rng.normal() - 0.5);
        } while (1 < Math.sqrt(b * b + c * c));
        allAs.push(b);
        allBs.push(c);
    }
}

function bang() {
    var indices = order.slice(0, take).sort(function (a, b) { return a - b });

    var as = [];
    var bs = [];
    for (let i = 0; i < take; i++) {
        as.push(allAs[indices[i]]);
        bs.push(allBs[indices[i]]);
    }
    outlet(1, as);
    outlet(0, bs);
}

function loadbang() {
    regenerate();
    bang();
}

let module = {};
export = {};

File: randomCs.java
import com.cycling74.max.*;
import java.security.SecureRandom;
import java.util.Arrays;

public class randomCs extends MaxObject {

    int maxLength = 64;

    final int[] order = new int[maxLength];

    randomCs() {
        order[0] = 0;
        order[1] = maxLength - 1;
        int pos = 2;
        search: for (double denominator = 2; denominator < 100; denominator++) {
            for (double numerator = 1; numerator < denominator; numerator++) {
                double fraction = numerator / denominator;
                int number = (int) Math.round(fraction * maxLength);
                if (!Arrays.stream(order).anyMatch(i -> i == number)) {
                    order[pos++] = number;
                    if (pos == maxLength) {
                        break search;
                    }
                }
            }
        }

        declareInlets(new int[] { DataTypes.INT, DataTypes.INT });
        declareOutlets(new int[] { DataTypes.LIST, DataTypes.LIST });
    }

    int seed = 0;
    int take = 16;
    SecureRandom rng;

    float[] allAs = new float[maxLength];
    float[] allBs = new float[maxLength];

    protected void inlet(int b) {
        int idx = getInlet();
        switch (idx) {
            case 0:
                take = b;
                break;
            case 1:
                seed = b;
                regenerate();
                break;
        }
        bang();
    }

    void regenerate() {
        rng = new SecureRandom(new byte[] { (byte) (seed - 128) });
        for (int i = 0; i < maxLength; i++) {
            float a, b;
            do {
                a = 2f * (rng.nextFloat() - 0.5f);
                b = 2f * (rng.nextFloat() - 0.5f);
            } while (1 < Math.sqrt(a * a + b * b));
            allAs[i] = a;
            allBs[i] = b;
        }
    }

    protected void bang() {
        int[] indices = new int[take];
        System.arraycopy(order, 0, indices, 0, take);
        Arrays.sort(indices);

        float[] as = new float[take];
        float[] bs = new float[take];
        for (int i = 0; i < take; i++) {
            as[i] = allAs[indices[i]];
            bs[i] = allBs[indices[i]];
        }
        outlet(1, as);
        outlet(0, bs);
    }
}
